// @polkadot/types/metadata/decorate/storage/index.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) { symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); } keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

// Copyright 2017-2021 @polkadot/types authors & contributors
// SPDX-License-Identifier: Apache-2.0
import { stringCamelCase, stringLowerFirst } from '@polkadot/util';
import { createFunction, createKeyRaw } from "./createFunction.js";
import { getStorage } from "./getStorage.js";
import { createRuntimeFunction } from "./util.js";
/** @internal */
// eslint-disable-next-line @typescript-eslint/no-unused-vars

export function decorateStorage(registry, {
  pallets
}, _metaVersion) {
  return pallets.reduce((result, moduleMetadata) => {
    if (moduleMetadata.storage.isNone) {
      return result;
    }

    const {
      name
    } = moduleMetadata;
    const section = stringCamelCase(name);
    const unwrapped = moduleMetadata.storage.unwrap();
    const prefix = unwrapped.prefix.toString(); // For access, we change the index names, i.e. System.Account -> system.account

    result[section] = unwrapped.items.reduce((newModule, meta) => {
      const method = meta.name.toString();
      newModule[stringLowerFirst(method)] = createFunction(registry, {
        meta,
        method,
        prefix,
        section
      }, {});
      return newModule;
    }, {
      palletVersion: createRuntimeFunction(name.toString(), 'palletVersion', createKeyRaw(registry, {
        method: ':__STORAGE_VERSION__:',
        prefix: name.toString()
      }, [], [], []), {
        docs: 'Returns the current pallet version from storage',
        type: 'u16'
      })(registry)
    });
    return result;
  }, _objectSpread({}, getStorage(registry)));
}
